File: Training/20141104/machine/machine.cpp
#include <fstream>
#include <string.h>

using namespace std;

ifstream fin("machine.in");
ofstream fout("machine.out");

const int MAX = 100+5;

int n, m, g[MAX][MAX];

int link[MAX], used[MAX];

int path(int i) {
   if (used[i]) return 0;
   used[i] = 1;
   int j;
   for (j = 1; j <= m; j++) if (g[i][j])
      if (!link[j] || path(link[j])) {
         link[j] = i;
         return 1;  
      }  
   return  0;   
}

main() {
   int i, j, k;
   fin >> n >> m >> k;

   memset(g, 0, sizeof(g));   
   for (i = 0; i < k; i++) {
      int temp, x, y;
      fin >> temp >> x >> y;
      if (x && y) g[x][y] = 1;  
   }
      
   memset(link, 0, sizeof(link));
   memset(used, 0, sizeof(used));
   
   for (i = 1; i <= n; i++)
      if (path(i))
         memset(used, 0, sizeof(used));
         
   int tot = 0;
   for (i = 1; i <= m; i++)
      if (link[i])
         tot++;
         
   fout << tot << endl;
   
   return 0;   
}

File: Training/20141104/machine/machine.pas
var
    g: array [1..100, 1..100] of Boolean;
    used: array [1..100] of Boolean;
    link:array[1..100] of longint;
    x, y, m, n, k, t, i: longint;
    tot: Longint;

function path(i: longint): Boolean;
var
    j: longint;
begin
    if used[i] then exit(false);
    used[i] := True;
    for j := 1 to m do
        if g[i][j] then
            if (link[j]=0) or path(link[j]) then
            begin
                link[j] := i;
                exit(True);
            end;
    exit(false);
end;

begin
    assign(input, 'machine1.in'); reset(input);
    assign(output, 'machine.out'); rewrite(output);

    read(n, m, k);
    for i := 1 to k do
    begin
        read(t, x, y);
        if (x = 0) or (y = 0) then continue;
        g[x][y] := True;
    end;

    for i := 1 to n do
    begin
        if path(i) then
            fillchar(used, sizeof(used), 0);    
    end;

    for i := 1 to m do
        if link[i]<>0 then inc(tot);
    writeln(tot);

    close(input); close(output);
end.
